### ISSUE TITLE
Failure to catch RuntimeError exception in assert_raises prior to python 3.6

### ISSUE THREAD

**Issue by gersner** [getslash/slash#501]
Consider the following example which fails to catch RuntimeError generated exception from within generators running under `slash.assert_raises`.

Tested on
- Python 3.5.2/Linux/x64 FAIL
- Python 3.6.0/Linux/x64 SUCCESS

```
import slash
import contextlib

@contextlib.contextmanager
def generator_with_exception(exception_class):
        yield 1
        raise exception_class("WAT")

class Example(slash.Test):
        def test_sanity(self):
                with slash.assert_raises(Exception):
                        with generator_with_exception(Exception):
                                pass

        def test_unexpected(self):
                with slash.assert_raises(RuntimeError):
                        with generator_with_exception(RuntimeError):
                                pass
```
I did not dig enough to find the root cause, but it probably has to do with the following fix in CPython.
https://github.com/python/cpython/commit/945b0ae5bffbfb8fb6949badf7b5c57b3ba6d9e7
The patch is currently on branches 3.5 (Probably will get into 3.5.3 which is now rc1) and 3.6.

Gersner.

**Comment by vmalloc** [comment]
@gersner this is interesting. Any idea about the behavior on 2.x?

**Comment by vmalloc** [comment]
@gersner this seems like a known bug in `contextmanager` that we recently worked around using a "regular" class context manager. I'm pushing a side branch, `hotfix/1.1.6`, please check if it works for you, and if it does I'll release it as a hotfix.

**Comment by gersner** [comment]
@vmalloc Thanks for the patch! We bumped to 3.6 which came out last week, so it's a bit hard to run the original test (which now uses 3.6 specific features).

I ran the sanity code with the patched Slash on Python (Ubuntu\x64) versions 2.7.12,3.5.2 and 3.6.0 and it looks OK.

I've also tested 2.7.12 without the patch (Slash 1.3.5) and it passes! The regression was probably introduced somewhere in 3.x.
